service/interpreter/stateExecutionCounter.go

```diff
@@ -69,55 +69,106 @@ func (e *StateExecutionCounter) CreateNextExecutionId(stateId string) string {
 }
 
 func (e *StateExecutionCounter) MarkStateIdExecutingIfNotYet(stateReqs []StateRequest) error {
+	config := e.configer.Get()
+
 	needsUpdateSA := false
 	numOfNew := 0
 	for _, sr := range stateReqs {
 		if sr.IsResumeRequest() {
 			continue
 		}
 		s := sr.GetStateStartRequest()
-		numOfNew++
-		e.stateIdCurrentlyExecutingCounts[s.StateId]++
-		if e.stateIdCurrentlyExecutingCounts[s.StateId] == 1 {
-			// first time the stateId show up
-			needsUpdateSA = true
+
+		if e.globalVersioner.IsAfterVersionOfExecutingStateIdMode() {
+			switch mode := config.GetExecutingStateIdMode(); mode {
+			case "DISABLED":
+				// do nothing
+			case "ENABLED_FOR_ALL":
+				if e.IncreaseStateIdCurrentlyExecutingCounts(s) {
+					needsUpdateSA = true
+				}
+			default: // "ENABLED_FOR_STATES_WITH_WAIT_UNTIL" or nil or unrecognized enum value
+				options := s.GetStateOptions()
+				if !options.GetSkipWaitUntil() {
+					if e.IncreaseStateIdCurrentlyExecutingCounts(s) {
+						needsUpdateSA = true
+					}
+				}
+			}
+		} else {
+			if !config.GetDisableSystemSearchAttribute() {
+				if e.IncreaseStateIdCurrentlyExecutingCounts(s) {
+					needsUpdateSA = true
+				}
+			}
 		}
+
+		numOfNew++
 	}
 	e.totalCurrentlyExecutingCount += numOfNew
+
 	if needsUpdateSA {
-		return e.updateStateIdSearchAttribute()
+		return e.UpdateStateIdSearchAttribute()
 	}
 	return nil
 }
 
+func (e *StateExecutionCounter) IncreaseStateIdCurrentlyExecutingCounts(s iwfidl.StateMovement) bool {
```

Comment:
In Golang, capitalizing first letter make it "public", otherwise it's "private" within the package.
Here in iwf, it is just a convention that we only capitalizing the methods that are being called outside of this file. So if `IncreaseStateIdCurrentlyExecutingCounts` is only being used within this file (by other methods), we would keep it private. This is to make the code a little more readable (moving them into different packages will be very hard because Golang doesn't allow circular dependency acorss packages)